File: majorityElement2.cpp

#include <iostream>
using namespace std;

#include <vector>
#include <map>

/*
	Intuition: Simple traverse looking for numbers which are greater than nums.size() / 3.

	RunTime: O(N)
	Space Complexity: O(N)
*/

vector<int> majorityElement(vector<int>& nums) 
    {
        map<int, int> myMap;
        vector<int> result;
        
        for (auto i: nums)
        {
            myMap[i]++;
        }
        
        for (auto i: myMap)
        {
            if (i.second > (nums.size() / 3))
            {
                result.push_back(i.first);
            }
        }
        
        return result;
    }


int main()
{
	vector<int> nums = {1,1,1,3,3,2,2,2};
	
	vector<int> answer = majorityElement(nums);

	for (auto i: answer)
	{
		cout << i << " ";
	}

	cout << endl;

	return 0;
}

File: majorityElement2.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        myMap = {}      #empty dictionary
        output = []     #empty vector

        for i in range(0, len(nums)):
            if nums[i] not in myMap:
                myMap[nums[i]] = 1
            elif nums[i] in myMap:
                myMap[nums[i]] = myMap[nums[i]] + 1
            #print(myMap[nums[i]])
            if (myMap[nums[i]] > (len(nums) / 3)):
                if (nums[i] not in output):
                    output.append(nums[i])
        
        return output
    
